src/Microsoft.Health.Dicom.Core/Features/Store/Entries/DicomInstanceEntryReaderForSinglePartRequest.cs

```diff
@@ -52,10 +62,110 @@ public async Task<IReadOnlyList<IDicomInstanceEntry>> ReadAsync(string contentTy
                     string.Format(CultureInfo.InvariantCulture, DicomCoreResource.UnsupportedContentType, contentType));
             }
 
-            Stream seekableStream = await _seekableStreamConverter.ConvertAsync(stream, cancellationToken);
+            Stream seekableStream;
+            using (Stream limitStream = new LimitStream(stream, _storeConfiguration.Value.MaxAllowedDicomFileSize))
+            {
+                seekableStream = await _seekableStreamConverter.ConvertAsync(limitStream, cancellationToken);
+            }
+
             dicomInstanceEntries.Add(new StreamOriginatedDicomInstanceEntry(seekableStream));
 
             return dicomInstanceEntries;
         }
+
+        private class LimitStream : Stream
+        {
+            private readonly Stream _stream;
+
+            private long _bytesLeft;
+
+            private readonly long _limit;
+
+            public override bool CanRead => _stream.CanRead;
+
+            public override bool CanSeek => false;
+
+            public override bool CanWrite => _stream.CanWrite;
```

Comment:
It may be best just to remove write capability like we did with `Seek`. Then we could call this stream something like `ReadOnlyLimitStream`